File: nim/triangle/Iteration1/triangle.nim

proc allSidesNonZero(triangle: array[3, int]): bool =
  triangle[0] > 0 and triangle[1] > 0 and triangle[2] > 0 

proc allLengthsValid(triangle: array[3, int]): bool =
  let side1 = triangle[0]
  let side2 = triangle[1]
  let side3 = triangle[2]

  side1 + side2 >= side3 and side1 + side3 >= side2 and side2 + side3 >= side1

proc isValidTriangle(triangle: array[3, int]): bool =
  allSidesNonZero(triangle) and allLengthsValid(triangle)

proc isEquilateral*(triangle: array[3, int]): bool =
  let side1 = triangle[0]
  let side2 = triangle[1]
  let side3 = triangle[2]

  isValidTriangle(triangle) and side1 == side2 and side1 == side3 and side2 == side3

proc isIsosceles*(triangle: array[3, int]): bool =
  let side1 = triangle[0]
  let side2 = triangle[1]
  let side3 = triangle[2]

  isValidTriangle(triangle) and (side1 == side2 or side1 == side3 or side2 == side3)

proc isScalene*(triangle: array[3, int]): bool =
  let side1 = triangle[0]
  let side2 = triangle[1]
  let side3 = triangle[2]

  isValidTriangle(triangle) and side1 != side2 and side1 != side3 and side2 != side3

File: zig/triangle/Iteration1/triangle.zig

pub const TriangleError = error{
    Invalid,
};

pub const Triangle = struct {
    a: f64,
    b: f64,
    c: f64,

    pub fn init(a: f64, b: f64, c: f64) TriangleError!Triangle {
        if (!isValid(a, b, c)) {
            return TriangleError.Invalid;
        }
    
        return Triangle {
            .a = a,
            .b = b,
            .c = c,
        };
    }

    fn isValid(a: f64, b: f64, c: f64) bool {
        return a > 0 and
               b > 0 and
               c > 0 and
               a + b >= c and
               b + c >= a and
               a + c >= b;
    }

    pub fn isEquilateral(self: Triangle) bool {
        return self.a == self.b and self.b == self.c and self.c == self.a;
    }

    pub fn isIsosceles(self: Triangle) bool {
        return self.a == self.b or self.b == self.c or self.c == self.a;
    }

    pub fn isScalene(self: Triangle) bool {
        return self.a != self.b and self.a != self.c and self.c != self.b;
    }
};
